docs/intro/from-ground-up.rst

```diff
@@ -498,151 +527,160 @@ to
+            self.crawl_state = crawl_state
+
+For each possible input a separate class needs to be defined, even if the
+data has the same format. For example, both ``ResponseData`` and
+``SplashResponseData`` may have the same ``url`` and ``html`` fields,
+but they can't be the same class, because they need to work as
+"markers" - tell frameworks if the html should be taken from HTTP
+response body or from Splash DOM snapshot.
+
+``CrawlState`` in the example above can be defined as a class with
+some specific properties, or maybe even
+as a ``class CrawlState(dict): pass`` - an important thing is that it is
+an unique type, and that we agree on what should be put into
+arguments annotated as ``CrawlState``.
+
+web-poet role
+=============
+
+How do you actually inspect the ``__init__`` method signature - e.g.
+if you're working on supporting ``web-poet`` page objects in some
+framework? ``web-poet`` itself doesn't provide any helpers for doing this.
+
+Use andi_ library. For example, scrapy-poet_ uses andi_.
+In addition to signature inspection, it also handles
+typing.Optional and typing.Union, and allows to create a build plan
+for dependency trees, indirect dependencies: that's allowed to annotate
+an argument as another :class:`~.Injectable` subclass.
 
-1. A list of possible inputs for the page objects. This helps with
+``web-poet`` is not using andi_ on its own; ``web-poet``'s role
+is to standardize things + provide some helpers to write the
+extraction code easier:
+
+1. Standardize a list of possible inputs for the page objects. This helps with
    reusability of extraction code across different environments. For example,
    if you want to support extraction from raw HTTP response bodies, you
    need to figure out how to populate :class:`~.ResponseData` in the
    given environment, and that's all.
 
-   Users are free to define their own inputs, but they may be less portable
-   across environments.
+   Users are free to define their own inputs (input types), but they
+   may be less portable across environments - which can be fine.
 
    Currently only :class:`~.ResponseData` is defined in web-poet.
 
-2. Interface for the Page Object itself. This allows to have a code which can
-   instantiate and use a Page Object without knowing about its
-   implementation upfront. ``web-poet`` requires you to use a base class,
-   and defines the semantics of ``to_item()`` method.
+2. Define an interface for the Page Object itself. This allows to
+   have a code which can instantiate and use a Page Object without knowing
+   about its implementation upfront. ``web-poet`` requires you to
+   use a base class (:class:`~.Injectable`), and defines the
+   semantics of ``to_item()`` method.
+
+
+Then, framework's role is to:
+
+1. Figure out which inputs a Page Object needs, likely using andi_ library.
+2. Create all the necessary inputs. For example, creating
+   :class:`~.ResponseData` instance may involve making an HTTP request.
```

Comment:
Would make sense to state `CrawlState` as another thing to be provided by the framework? I'm saying that because it is a little bit obscure where `CrawlState` comes from.

Reply:
Also, we are not saying anywhere that any Injectable can be used as a dependency. We should say it somewhere.

Reply:
> Also, we are not saying anywhere that any Injectable can be used as a dependency. We should say it somewhere.

There is a passing note about it:

> In addition to signature inspection, it also handles
typing.Optional and typing.Union, and allows to create a build plan
for dependency trees, indirect dependencies: that's allowed to annotate
an argument as another :class:`~.Injectable` subclass.

These points deserve a dedicated section: how to use Optional and Union, how to use Injectable dependencies. I would prefer to have a separate task for this though.

Reply:
Makes sense.